lib/easybill_rest_client/request.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+require 'easybill_rest_client/request_logger'
+require 'easybill_rest_client/retry_on'
+
+module EasybillRestClient
+  class Request
+    BASE_URL = 'https://api.easybill.de/rest/v1'
+    USERNAME = 'rest-api@easybill.de'
+    SUPPORTED_METHODS = %i(get post put delete).freeze
+    OPEN_TIMEOUT = 5
+
+    def initialize(opts = {})
+      unless SUPPORTED_METHODS.include?(opts.fetch(:method).to_sym)
+        raise ArgumentError, "Unsupported HTTP method '#{method}'"
+      end
+
+      @api_key = opts.fetch(:api_key)
+      @method = opts.fetch(:method)
+      @endpoint = opts.fetch(:endpoint)
+      @params = opts.fetch(:params)
+      @logger = opts.fetch(:logger)
+      @tries = opts.fetch(:tries)
+      @retry_cool_off_time = opts.fetch(:retry_cool_off_time)
+    end
+
+    def run
+      request_logger.info("#{method.to_s.upcase} #{endpoint} #{request_details}")
+      retry_on(TooManyRequests) do
+        retry_on(Net::OpenTimeout) do
+          response = perform_request
+          raise TooManyRequests if response.is_a?(Net::HTTPTooManyRequests)
+          response
+        end
+      end
+    end
+
+    def request_details
+      body_allowed? ? request.body : uri.query
+    end
+
+    private
+
+    attr_reader :api_key, :method, :endpoint, :params, :logger, :request_logger, :retry_on, :tries,
+      :retry_cool_off_time
+
+    def perform_request
+      http_opts = { use_ssl: uri.scheme == 'https', open_timeout: OPEN_TIMEOUT }
+      Net::HTTP.start(uri.host, uri.port, http_opts) do |http|
+        http.request(request)
+      end
+    end
+
+    def uri
+      return @uri if @uri
+      @uri = URI.parse(BASE_URL)
+      @uri.path << endpoint
+      @uri.query = URI.encode_www_form(comma_separate_arrays(params.dup)) unless body_allowed?
+      @uri
+    end
+
+    def request
+      return @request if @request
+      @request = request_class.new(uri)
+      @request.body = params.reject { |_k, v| v.nil? }.to_json if body_allowed?
+      @request.basic_auth(USERNAME, api_key)
+      @request
+    end
+
+    def request_class
+      Net::HTTP.const_get(method.to_s.capitalize)
+    end
+
+    def body_allowed?
+      request_class::REQUEST_HAS_BODY
+    end
+
+    def comma_separate_arrays(params)
+      params.map { |k, v| [k, v.is_a?(Array) ? v.join(',') : v] }.to_h
+    end
+
+    def retry_on(exception, &block)
+      @retry_on ||= RetryOn.new(request_logger, tries, retry_cool_off_time)
+      @retry_on.retry_on(exception, &block)
+    end
+
+    def request_logger
+      @request_logger ||= RequestLogger.new(logger: logger, request_id: request_id)
```

Comment:
> So what's stopping you? Just put in a `logger.info("[#{request.id}] starting request #{request.params}")` or similar in there somewhere

Nothing is stopping me, but I'd have to put the `[easybill-rest-client]` string in there as well. Won't you agree, that formatting the log line in one place instead of several is better (DRY) and that it's the responsibility of the logger, and not the request object nor the exceptions' to do so (SRP)?

> Now you have to create a new logger instance for each request.

So?

> It's also a lot more code that's more difficult to understand.

To that I'll just say _I disagree_ and by that provide as much of an argument as you do.

When reviewing, keep in mind that it costs time to write feedback, argue about it, _make up actual example code to put into a comment_, change the code (or not), and, last but not least, the finance department's when they wait for the page to load, only to have an error message waiting for them when they get back to it, and whether the change you are proposing is actually worth all this.

Unless you have any other feedback, I'll consider this good to go and move on.